### PR TITLE
Fix signup with-theme flow for Marketplace themes and other improvements

### PR DESCRIPTION
**Pull request by BogdanUngureanu**
Fix the with-theme flow for Marketplace themes and filter the plans based on the type of the theme:
- For Marketplace and DotOrg themes we will now show just `Business` and `eCommerce` plans.
- For Premium themes, we will now show `Premium`, `Business` and `eCommerce` plans - this is needed because the user will be redirected to designSetup step where they'll be asked yet again to purchase `Premium`.

In both cases, we still display the `start with free` link in the header. I think we need to allow users an escape hatch, in case they don't want to purchase a qualified plan for the theme.

### TODO in a separate PR
You'll notice that there are some discrepancies here for sites that select an incompatible plan (e.g. free for premium themes): 
- Marketplace themes, Users are redirected to DesignSetup theme list where they can select a different theme
- DotORG themes, users are redirected to /marketplace/install/ where an upgrade nudge is displayed asking for an upgrade to Business
- Premium themes, users are redirected to a dedicated /DesignSetup page for a given theme, where they are asked again to upgrade.

We should come up with a flow consistent for all flows, as much as possible, given the current limitations (e.g. dot-org and Marketplace themes don't work on DesignSetup).

Related to https://github.com/Automattic/wp-calypso/issues/75335

## Proposed Changes

* Add a new plans step that handles the with-theme flows, especially the Marketplace themes.

## Testing Instructions

* Go incognito to WordPress.com/themes
* Right-click on the Pick design button and copy the /start/with-theme link
* Open a browser window where you are already logged-in and go to that URL.
* Make sure that for Marketplace themes:
* * we only display the Business and eCommerce plans
* * After selecting one of the plans, you are redirected to checkout with the plan added and with the theme subscription selected
* * Make sure that when you select a yearly/monthly plan subscription, the theme subscription has the same recurrence
* * Make sure that after you are purchasing the theme and plan, you are now getting an AT site with the theme installed.
* * Make sure that when you click on the `or Start with free`, you are redirected to the DesignSetup theme list, where you can select another theme
* Make sure that for Premium themes we only display Premium, Business and eCommerce plans
* Make sure that for DotORG themes we only display Business and eCommerce plans
## Pre-merge Checklist

- [ ] Has the general commit checklist been followed? (PCYsg-hS-p2)
- [ ] [Have you written new tests](https://wpcalypso.wordpress.com/devdocs/docs/testing/index.md) for your changes?
- [ ] Have you tested the feature in Simple (P9HQHe-k8-p2), Atomic (P9HQHe-jW-p2), and self-hosted Jetpack sites (PCYsg-g6b-p2)?
- [ ] Have you checked for TypeScript, React or other console errors?
- [ ] Have you used memoizing on expensive computations? More info in [Memoizing with create-selector](https://github.com/Automattic/wp-calypso/blob/trunk/packages/state-utils/src/create-selector/README.md) and [Using memoizing selectors](https://react-redux.js.org/api/hooks#using-memoizing-selectors) and [Our Approach to Data](https://github.com/Automattic/wp-calypso/blob/trunk/docs/our-approach-to-data.md)
- [ ] Have we added the "[Status] String Freeze" label as soon as any new strings were ready for translation (p4TIVU-5Jq-p2)?
- [ ] For changes affecting Jetpack: Have we added the "[Status] Needs Privacy Updates" label if this pull request changes what data or activity we track or use (p4TIVU-ajp-p2)?

### CODE REVIEW

**Comment by mmtr** [comment]
> For Marketplace and DotOrg themes we will now show just Business and eCommerce plans.

Can we also ensure that WooCommerce themes only show Business and eCommerce plans?

Screenshot 2023-04-27 at 13 35 42

**Comment by mmtr** [comment]
> Marketplace themes, Users are redirected to DesignSetup theme list where they can select a different theme

I think it might be safer/less confusing if, for this scenario (when a user selects a Marketplace theme but doesn't purchase the required plan), we redirect users to `/theme/:theme/:site`.

**Comment by mmtr** [comment]
> Make sure that after you are purchasing the theme and plan, you are now getting an AT site with the theme installed

The theme gets installed as expected, but I see an "Activate this design" button that apparently does nothing?

https://user-images.githubusercontent.com/1233880/234853062-7c08f3a9-c5e0-4441-b110-426dfac4635d.mov

It might be unrelated to this PR though.

File: client/lib/signup/step-actions/index.js [hunk 1]
```diff
@@ -486,6 +496,102 @@ export function setIntentOnSite( callback, { siteSlug, intent } ) {
...
+				dependencies.siteSlug,
+				cartItem
+			).then( () => {} );
+		} else {
+			addPlanToCart( callback, dependencies, stepProvidedItems, reduxStore );
+		}
+	} );
+}
+
+const isLoadingCart = ( isLoading ) => ( dispatch ) => {
```
Comment (mmtr) [inline client/lib/signup/step-actions/index.js]: Nitpick (non-blocking): The name of this action resembles a selector. Maybe `setIsLoadingCard` is clearer?

**Review by mmtr — COMMENTED** [review]
This works great, nice job! I left a few comments for consideration – let me know what do you think about them. If needed, we can address them in follow-ups

**Comment by BogdanUngureanu** [comment]
> I think it might be safer/less confusing if, for this scenario (when a user selects a Marketplace theme but doesn't purchase the required plan), we redirect users to /theme/:theme/:site

Not sure about that because the user opted for `or a free site`. Redirecting them back to the theme doesn't bring any value to them; they've already decided that they don't want that theme.

I think it makes sense to point them to the design picker because they'll have to select another theme - and (partially) start the flow again.

**Comment by BogdanUngureanu** [comment]
> The theme gets installed as expected, but I see an "Activate this design" button that apparently does nothing?

This sounds like a Marketplace issue; the with-theme Marketplace implementation just points them to the checkout page, and after that the Marketplace implementation kicks in. cc @Automattic/bespin

**Comment by mmtr** [comment]
> Redirecting them back to the theme doesn't bring any value to them; 

Is not that what we do for premium themes? I don't think clicking on "start with free" necessarily means they don't want that theme anymore. They might think they can get the theme they selected with a free plan. And redirecting them back to the theme details page gives them another opportunity to realize that (1) the theme has not been activated and (2) they need to purchase a plan.

**Comment by BogdanUngureanu** [comment]
> > Redirecting them back to the theme doesn't bring any value to them;
> 
> Is not that what we do for premium themes? I don't think clicking on "start with free" necessarily means they don't want that theme anymore. They might think they can get the theme they selected with a free plan. And redirecting them back to the theme details page gives them another opportunity to realize that (1) the theme has not been activated and (2) they need to purchase a plan.

Yes, but I think we should change it, or at least give them the option for them to go to Design Picker, otherwise they'll be stuck on the page. 🤔 

> And redirecting them back to the theme details page gives them another opportunity to realize that (1) the theme has not been activated and (2) they need to purchase a plan.

In this case, they'll have to click on the upgrade button in the theme details, and they are redirected to checkout. We don't inform them upfront what are the requirements and prices. 

What if we rename the link to something like `Or start with another theme & free site"? 🤔 

Maybe we should also add another message like "XXX(theme) requires a paid plan. Pick one that's right for you and unlock feature that help you grow. Or start with another theme & free site".

**Comment by mmtr** [comment]
> otherwise they'll be stuck on the page

Good point, didn't consider this! Let's leave it as is for now, and revisit it once we can get some feedback from designers.

**Review by mmtr — APPROVED** [review]

**Pull request merged**